examples/text/plot_hashing_vs_dict_vectorizer.py

```diff
@@ -3,109 +3,343 @@
 
+import re
 
-def tokens(doc):
+
+def tokenize(doc):
     """Extract tokens from doc.
 
-    This uses a simple regex to break strings into tokens. For a more
-    principled approach, see CountVectorizer or TfidfVectorizer.
+    This uses a simple regex that matches word characters to break strings
+    into tokens. For a more principled approach, see CountVectorizer or
+    TfidfVectorizer.
     """
     return (tok.lower() for tok in re.findall(r"\w+", doc))
 
 
+list(tokenize("This is a simple example, isn't it?"))
+
+# %%
+# We define an additional function that transforms a document into a frequency
+# dictionary to be used by the vectorizers.
+
+from collections import defaultdict
+
+
 def token_freqs(doc):
-    """Extract a dict mapping tokens from doc to their frequencies."""
+    """Extract a dict mapping tokens from doc to their frequencies.
+
+    Count the number of occurrences of all tokens of each document.
+    """
+
     freq = defaultdict(int)
-    for tok in tokens(doc):
+    for tok in tokenize(doc):
         freq[tok] += 1
     return freq
 
 
-categories = [
-    "alt.atheism",
-    "comp.graphics",
-    "comp.sys.ibm.pc.hardware",
-    "misc.forsale",
-    "rec.autos",
-    "sci.space",
-    "talk.religion.misc",
-]
-# Uncomment the following line to use a larger set (11k+ documents)
-# categories = None
-
-print(__doc__)
-print("Usage: %s [n_features_for_hashing]" % sys.argv[0])
-print("    The default number of features is 2**18.")
-print()
+token_freqs("That is one example, but this is another one")
 
-try:
-    n_features = int(sys.argv[1])
-except IndexError:
-    n_features = 2**18
-except ValueError:
-    print("not a valid number of features: %r" % sys.argv[1])
-    sys.exit(1)
+# %%
+# DictVectorizer
+# --------------
+#
+# First we benchmark the :func:`~sklearn.feature_extraction.DictVectorizer`,
+# then we compare it to :func:`~sklearn.feature_extraction.FeatureHasher` as
+# both of them receive dictionaries as input.
 
+from time import time
+from sklearn.feature_extraction import DictVectorizer
 
-print("Loading 20 newsgroups training data")
-raw_data, _ = fetch_20newsgroups(subset="train", categories=categories, return_X_y=True)
-data_size_mb = sum(len(s.encode("utf-8")) for s in raw_data) / 1e6
-print("%d documents - %0.3fMB" % (len(raw_data), data_size_mb))
-print()
+dict_count_vectorizers = defaultdict(list)
 
-print("DictVectorizer")
 t0 = time()
 vectorizer = DictVectorizer()
 vectorizer.fit_transform(token_freqs(d) for d in raw_data)
 duration = time() - t0
-print("done in %fs at %0.3fMB/s" % (duration, data_size_mb / duration))
-print("Found %d unique terms" % len(vectorizer.get_feature_names_out()))
-print()
+dict_count_vectorizers["vectorizer"].append(vectorizer.__class__.__name__)
+dict_count_vectorizers["speed"].append(data_size_mb / duration)
+print(f"done in {duration:.3f}s at {data_size_mb / duration:.3f}MB/s")
+print(f"Found {len(vectorizer.get_feature_names_out())} unique terms")
+
```

Comment:
To better introduce the motivation for feature hashing, let's highlight that the mapping is explicitly stored in memory as a Python dict.

```suggestion
print(f"Found {len(vectorizer.get_feature_names_out())} unique terms")

# %%
# The actual mapping from text to token to column index is explicitly stored in the
# `.vocabulary_` attribute which is a potentially very large Python dictionary:
type(vectorizer.vocabulary_)

# %%
len(vectorizer.vocabulary_)

# %%
vectorizer.vocabulary_["example"]
```